mountpoint-s3/src/checksums/block.rs

```diff
@@ -0,0 +1,213 @@
+use bytes::{Bytes, BytesMut};
+use mountpoint_s3_crt::checksums::crc32c::{self, Crc32c};
+
+use crate::checksums::{bytes::ChecksummedBytes, combine_checksums, IntegrityError};
+
+/// A `ChecksummedBlock` is a bytes buffer that carries its checksum.
+/// The implementation guarantees that its integrity will be validated when data is accessed.
+#[derive(Debug, Clone)]
+pub struct ChecksummedBlock {
+    bytes: Bytes,
+    checksum: Crc32c,
+}
+
+impl ChecksummedBlock {
+    pub fn new(bytes: Bytes, checksum: Crc32c) -> Self {
+        Self { bytes, checksum }
+    }
+
+    /// Create `ChecksummedBlock` from `Bytes`, calculating its checksum.
+    pub fn from_bytes(bytes: Bytes) -> Self {
+        let checksum = crc32c::checksum(&bytes);
+        Self::new(bytes, checksum)
+    }
+
+    /// Convert the `ChecksummedBlock` into `Bytes`, data integrity will be validated before converting.
+    ///
+    /// Return `IntegrityError` on data corruption.
+    pub fn into_bytes(self) -> Result<Bytes, IntegrityError> {
+        self.validate()?;
+
+        Ok(self.bytes)
+    }
+
+    /// Convert into a `ChecksummedBytes`.
+    pub fn into_checksummed_bytes(self) -> ChecksummedBytes {
+        ChecksummedBytes::new(self.bytes, self.checksum)
+    }
+
+    /// Returns the number of bytes contained in this `ChecksummedBlock`.
+    pub fn len(&self) -> usize {
+        self.bytes.len()
+    }
+
+    /// Returns true if the `ChecksummedBlock` has a length of 0.
+    pub fn is_empty(&self) -> bool {
+        self.bytes.is_empty()
+    }
+
+    /// Append the given bytes to current `ChecksummedBlock`.
+    pub fn extend(&mut self, extend: ChecksummedBlock) {
+        if self.is_empty() {
+            *self = extend;
+            return;
+        }
+        if extend.is_empty() {
+            return;
+        }
+
+        let total_len = self.bytes.len() + extend.len();
+        let mut bytes_mut = BytesMut::with_capacity(total_len);
+        bytes_mut.extend_from_slice(&self.bytes);
+        bytes_mut.extend_from_slice(&extend.bytes);
+        let new_bytes = bytes_mut.freeze();
+        let new_checksum = combine_checksums(self.checksum, extend.checksum, extend.len());
+        *self = ChecksummedBlock {
+            bytes: new_bytes,
+            checksum: new_checksum,
+        };
+    }
+
+    /// Validate data integrity in this `ChecksummedBlock`.
+    ///
+    /// Return `IntegrityError` on data corruption.
+    pub fn validate(&self) -> Result<(), IntegrityError> {
```

Comment:
Do we ever use this as public API? If not, might be better to make it private, since it kinda invites time-of-check/time-of-use problems.